File: metacasa-ios/MetaCasa/Core/Money.swift
import Foundation

/// Formatter centralizado de dinero + parsing robusto.
///
/// Usa `FormatStyle` moderno de Swift (iOS 15+). Respeta el locale efectivo
/// de la app (override del usuario > locale del sistema) y el currency code
/// del hogar actual.
///
/// Ejemplos:
///   Money.format(1234.56, currency: "ARS", locale: Locale(identifier: "es_AR"))
///     → "$ 1.234,56"
///   Money.format(1234.56, currency: "USD", locale: Locale(identifier: "en_US"))
///     → "$1,234.56"
///   Money.format(1234.56, currency: "EUR", locale: Locale(identifier: "es_ES"))
///     → "1.234,56 €"
enum Money {
    /// Estilo de presentación.
    enum Style {
        /// Sin decimales (estilo LatAm para montos grandes, UX limpia). `$ 1.234`
        case compact
        /// Con 2 decimales (exacto). `$ 1.234,56`
        case precise
        /// Auto: compact si el monto es entero, precise si tiene fracción.
        case auto
        /// Notación compacta con sufijos (K/M/B) según locale. `$4,4 M` / `$15 K`.
        /// Útil para celdas chicas (calendar, widgets, charts).
        case abbreviated
    }

    /// Formatea un monto como moneda según locale + código ISO 4217.
    static func format(
        _ amount: Decimal,
        currency: String,
        style: Style = .compact,
        showSign: Bool = false,
        locale: Locale? = nil
    ) -> String {
        let loc = locale ?? Self.currentLocale()

        let base: String
        if case .abbreviated = style {
            // Notación abreviada con sufijos K/M/B/T (universal, no requiere iOS 18).
            // Ej: 4_399_630 → "$4,4M"; 15_000 → "$15K"; 999 → "$999".
            let abs = (amount as NSDecimalNumber).doubleValue.magnitude
            let (divisor, suffix): (Decimal, String) = {
                switch abs {
                case 1_000_000_000_000...: return (1_000_000_000_000, "T")
                case 1_000_000_000...:     return (1_000_000_000, "B")
                case 1_000_000...:         return (1_000_000, "M")
                case 1_000...:             return (1_000, "K")
                default:                   return (1, "")
                }
            }()
            let scaled = amount / divisor
            let fracDigits = suffix.isEmpty ? 0 : (abs.truncatingRemainder(dividingBy: Double(truncating: divisor as NSNumber)) > 0 ? 1 : 0)
            let formatStyle = Decimal.FormatStyle.Currency(code: currency, locale: loc)
                .precision(.fractionLength(fracDigits...fracDigits))
            base = scaled.formatted(formatStyle) + suffix
        } else {
            // Determinar fracción según estilo
            let (minFrac, maxFrac): (Int, Int) = {
                switch style {
                case .compact: return (0, 0)
                case .precise: return (2, 2)
                case .auto:    return hasFraction(amount) ? (2, 2) : (0, 0)
                case .abbreviated: return (0, 0) // unreachable
                }
            }()
            let formatStyle = Decimal.FormatStyle.Currency(code: currency, locale: loc)
                .precision(.fractionLength(minFrac...maxFrac))
            base = amount.formatted(formatStyle)
        }

        // Agregar signo '+' si corresponde (para ingresos/aportes positivos en UIs)
        if showSign && amount > 0 {
            return "+\(base)"
        }
        return base
    }

    /// Formatea un número puro (sin símbolo de moneda) respetando el locale.
    /// Útil para inputs, totales internos, etc.
    static func formatNumber(
        _ value: Decimal,
        fractionDigits: Int = 0,
        locale: Locale? = nil
    ) -> String {
        let loc = locale ?? Self.currentLocale()
        return value.formatted(
            .number
                .locale(loc)
                .precision(.fractionLength(fractionDigits...fractionDigits))
                .grouping(.automatic)
        )
    }

    /// Parsea un string con formato de moneda local al Decimal correspondiente.
    /// Tolera símbolos ($, US$, €, R$, ARS, USD), separadores de miles y decimales
    /// en ambas convenciones (es-AR `1.234,56` y en-US `1,234.56`).
    ///
    /// **Orden de estrategias**:
    /// 1. Heurística basada en el ÚLTIMO separador (más confiable — no depende
    ///    del locale y maneja ambos formatos correctamente). Ej: `-25000.00` →
    ///    25000 exacto. `1.234,56` → 1234.56. `1,234.56` → 1234.56.
    /// 2. NumberFormatter con el locale del caller (para strings con símbolos
    ///    raros que la heurística no capta).
    /// 3. Decimal(string:) — parser C-locale como último recurso.
    static func parse(_ input: String, locale: Locale? = nil) -> Decimal? {
        let loc = locale ?? Self.currentLocale()

        let stripped = input
            .replacingOccurrences(of: "$", with: "")
            .replacingOccurrences(of: "€", with: "")
            .replacingOccurrences(of: "U$S", with: "")
            .replacingOccurrences(of: "US$", with: "")
            .replacingOccurrences(of: "R$", with: "")
            .replacingOccurrences(of: "ARS", with: "")
            .replacingOccurrences(of: "USD", with: "")
            .replacingOccurrences(of: "EUR", with: "")
            .replacingOccurrences(of: "BRL", with: "")
            .replacingOccurrences(of: "\u{00A0}", with: "") // NBSP
            .trimmingCharacters(in: .whitespaces)

        // 1. Heurística — más robusta para CSV y entradas variadas.
        if let h = parseHeuristic(stripped) {
            return h
        }

        // 2. NumberFormatter con locale.
        let nf = NumberFormatter()
        nf.numberStyle = .decimal
        nf.locale = loc
        if let n = nf.number(from: stripped) {
            return n.decimalValue
        }

        // 3. Último recurso: parser C-locale (punto como decimal).
        return Decimal(string: stripped)
    }

    /// Heurística: el ÚLTIMO separador (. o ,) es el decimal; los anteriores son miles.
    private static func parseHeuristic(_ s: String) -> Decimal? {
        let lastDot = s.lastIndex(of: ".")
        let lastComma = s.lastIndex(of: ",")
        let decimalIndex: String.Index?
        switch (lastDot, lastComma) {
        case (nil, nil):          decimalIndex = nil
        case (let d?, nil):       decimalIndex = d
        case (nil, let c?):       decimalIndex = c
        case (let d?, let c?):    decimalIndex = d > c ? d : c
        }

        var normalized = s
        if let idx = decimalIndex {
            let intPart = s[..<idx]
                .replacingOccurrences(of: ".", with: "")
                .replacingOccurrences(of: ",", with: "")
            let fracPart = s[s.index(after: idx)...]
            normalized = "\(intPart).\(fracPart)"
        } else {
            normalized = s
                .replacingOccurrences(of: ".", with: "")
                .replacingOccurrences(of: ",", with: "")
        }
        return Decimal(string: normalized)
    }

    private static func hasFraction(_ amount: Decimal) -> Bool {
        var rounded = Decimal()
        var value = amount
        NSDecimalRound(&rounded, &value, 0, .plain)
        return amount != rounded
    }

    /// Locale efectivo global (lee override del usuario o cae a system).
    /// Nonisolated — seguro desde cualquier contexto.
    private static func currentLocale() -> Locale {
        AppLocaleStorage.effectiveLocale
    }
}

File: metacasa-flutter/lib/core/utils/money.dart
import 'package:decimal/decimal.dart';
import 'package:intl/intl.dart';

/// Estilo de presentación del monto. Espejo de `Money.Style` en iOS
/// (`Core/Money.swift`).
enum MoneyStyle {
  /// Sin decimales (estilo LatAm para montos grandes, UX limpia). `$ 1.234`
  compact,

  /// Con 2 decimales (exacto). `$ 1.234,56`
  precise,

  /// Auto: compact si el monto es entero, precise si tiene fracción.
  auto,

  /// Notación compacta con sufijos (K/M/B/T) según locale. `$4,4M` / `$15K`.
  /// Útil para celdas chicas (calendar, widgets, charts).
  abbreviated,
}

/// Formatter centralizado de dinero + parsing robusto.
///
/// Port del `enum Money` de iOS (`Core/Money.swift`). En lugar del
/// `Decimal.FormatStyle.Currency` de Swift usamos `NumberFormat` de `intl`
/// para el agrupamiento/decimales según locale, combinado con el símbolo
/// custom del [_symbols] (catálogo `CurrenciesCatalog.swift`).
///
/// Ejemplos:
///   Money.format(Decimal.parse('1234.56'), currencyCode: 'ARS', locale: 'es_AR')
///     → "$ 1.234"        (compact, 0 decimales)
///   Money.format(Decimal.parse('1234.56'), currencyCode: 'USD',
///                style: MoneyStyle.precise, locale: 'en_US')
///     → "U$S 1,234.56"
///   Money.format(Decimal.parse('1234.56'), currencyCode: 'EUR',
///                style: MoneyStyle.precise, locale: 'es_ES')
///     → "1.234,56 €"      (símbolo como sufijo)
class Money {
  const Money._();

  /// Símbolos custom por código ISO 4217 — port 1:1 de `CurrenciesCatalog.all`
  /// (`Core/CurrenciesCatalog.swift`). Difieren de los símbolos default de
  /// `intl` (ej. USD = `U$S`, no `$`).
  static const Map<String, String> _symbols = <String, String>{
    // América Latina
    'ARS': r'$',
    'BRL': r'R$',
    'CLP': r'CL$',
    'COP': r'CO$',
    'MXN': r'MX$',
    'PEN': 'S/.',
    'UYU': r'$U',
    'PYG': '₲',
    'BOB': 'Bs.',
    'VES': 'Bs.S',
    // Norte América
    'USD': r'U$S',
    'CAD': r'CA$',
    // Europa
    'EUR': '€',
    'GBP': '£',
    'CHF': 'Fr.',
    'SEK': 'kr',
    'NOK': 'kr',
    'DKK': 'kr',
    'PLN': 'zł',
    // Asia · Pacífico
    'JPY': '¥',
    'CNY': r'CN¥',
    'KRW': '₩',
    'INR': '₹',
    'AUD': r'A$',
    'NZD': r'NZ$',
    'SGD': r'S$',
    'HKD': r'HK$',
    // Medio Oriente · África
    'AED': 'د.إ',
    'ILS': '₪',
    'ZAR': 'R',
  };

  /// Monedas cuyo símbolo va como **sufijo** (`1.234,56 €`) en lugar de prefijo.
  /// El resto usa prefijo con espacio (`$ 1.234`), siguiendo la convención
  /// LatAm de la PWA/iOS.
  static const Set<String> _suffixCurrencies = <String>{
    'EUR',
    'SEK',
    'NOK',
    'DKK',
    'PLN',
    'CHF',
  };

  /// Símbolo custom para [code] (mayúsculas). Cae al propio código si no está
  /// en el catálogo.
  static String symbolFor(String code) =>
      _symbols[code.toUpperCase()] ?? code.toUpperCase();

  /// Formatea un monto como moneda según [locale] + código ISO 4217.
  ///
  /// - [style]: ver [MoneyStyle].
  /// - [showSign]: antepone `+` para montos positivos (ingresos/aportes).
  /// - [locale]: identificador ICU (`es_AR`, `en_US`, `pt_BR`, …).
  static String format(
    Decimal value, {
    required String currencyCode,
    MoneyStyle style = MoneyStyle.compact,
    bool showSign = false,
    String locale = 'es_AR',
  }) {
    final String symbol = symbolFor(currencyCode);
    final String body;

    if (style == MoneyStyle.abbreviated) {
      body = _abbreviated(value, locale: locale);
    } else {
      final int fractionDigits = switch (style) {
        MoneyStyle.compact => 0,
        MoneyStyle.precise => 2,
        MoneyStyle.auto => value.isInteger ? 0 : 2,
        MoneyStyle.abbreviated => 0, // inalcanzable
      };
      body = _grouped(value, fractionDigits: fractionDigits, locale: locale);
    }

    final String formatted = _withSymbol(body, symbol, currencyCode);

    // Signo '+' para positivos (no afecta a 0 ni a negativos).
    if (showSign && value > Decimal.zero) {
      return '+$formatted';
    }
    return formatted;
  }

  /// Combina cuerpo numérico + símbolo respetando prefijo/sufijo por moneda.
  static String _withSymbol(String body, String symbol, String code) {
    if (_suffixCurrencies.contains(code.toUpperCase())) {
      return '$body $symbol';
    }
    return '$symbol $body';
  }

  /// Cuerpo numérico agrupado según locale, con dígitos fraccionarios fijos.
  static String _grouped(
    Decimal value, {
    required int fractionDigits,
    required String locale,
  }) {
    final NumberFormat nf = NumberFormat.decimalPattern(locale)
      ..minimumFractionDigits = fractionDigits
      ..maximumFractionDigits = fractionDigits;
    // Redondeamos en Decimal y pasamos por double sólo para el agrupamiento
    // visual (la fuente de verdad sigue siendo Decimal).
    final Decimal rounded = value.round(scale: fractionDigits);
    return nf.format(rounded.toDouble());
  }

  /// Notación abreviada con sufijos K/M/B/T (universal, no depende del SO).
  /// Ej: 4_399_630 → "$4,4M"; 15_000 → "$15K"; 999 → "$999".
  /// Port de la rama `.abbreviated` de `Money.format` en iOS.
  static String _abbreviated(Decimal value, {required String locale}) {
    final double abs = value.toDouble().abs();
    final (Decimal divisor, String suffix) = switch (abs) {
      >= 1000000000000 => (Decimal.fromInt(1000000000000), 'T'),
      >= 1000000000 => (Decimal.fromInt(1000000000), 'B'),
      >= 1000000 => (Decimal.fromInt(1000000), 'M'),
      >= 1000 => (Decimal.fromInt(1000), 'K'),
      _ => (Decimal.one, ''),
    };
    final Decimal scaled =
        (value / divisor).toDecimal(scaleOnInfinitePrecision: 4);
    // 1 decimal si hay resto al dividir, 0 si es exacto (igual que iOS).
    final bool hasRemainder =
        suffix.isNotEmpty && abs % divisor.toDouble() != 0;
    final int fracDigits = hasRemainder ? 1 : 0;
    final String numberPart =
        _grouped(scaled, fractionDigits: fracDigits, locale: locale);
    return '$numberPart$suffix';
  }

  /// Parsea un string con formato de moneda local al `Decimal` correspondiente.
  /// Tolera símbolos ($, U$S, €, R$, ARS, USD…), separadores de miles y
  /// decimales en ambas convenciones (es-AR `1.234,56` y en-US `1,234.56`).
  ///
  /// Estrategia (port de `Money.parse` en iOS): heurística basada en el
  /// ÚLTIMO separador (`.` o `,`) = decimal; los anteriores = miles.
  static Decimal? parse(String input) {
    var stripped = input;
    for (final String token in const <String>[
      r'U$S',
      r'US$',
      r'R$',
      r'CL$',
      r'CO$',
      r'MX$',
      r'CA$',
      r'NZ$',
      r'HK$',
      r'S$',
      r'A$',
      r'CN¥',
      r'$U',
      r'$',
      '€',
      '£',
      '¥',
      '₲',
      '₩',
      '₹',
      '₪',
      'Bs.S',
      'Bs.',
      'Fr.',
      'S/.',
      'zł',
      'kr',
      'ARS',
      'USD',
      'EUR',
      'BRL',
      ' ', // NBSP
    ]) {
      stripped = stripped.replaceAll(token, '');
    }
    stripped = stripped.trim();
    if (stripped.isEmpty) return null;

    return _parseHeuristic(stripped) ?? Decimal.tryParse(stripped);
  }

  /// Heurística: el último separador es el decimal; el resto, miles.
  static Decimal? _parseHeuristic(String s) {
    final int lastDot = s.lastIndexOf('.');
    final int lastComma = s.lastIndexOf(',');
    final int decimalIndex = (lastDot < 0 && lastComma < 0)
        ? -1
        : (lastDot > lastComma ? lastDot : lastComma);

    final String normalized;
    if (decimalIndex >= 0) {
      final String intPart =
          s.substring(0, decimalIndex).replaceAll('.', '').replaceAll(',', '');
      final String fracPart = s.substring(decimalIndex + 1);
      normalized = '$intPart.$fracPart';
    } else {
      normalized = s.replaceAll('.', '').replaceAll(',', '');
    }
    return Decimal.tryParse(normalized);
  }
}
